lua-api-crates/table-funcs/src/lib.rs

```diff
@@ -0,0 +1,205 @@
+            }
+        } else {
+            tbl.set(key, value).map_err(mlua::Error::external)?;
+        }
+    }
+    Ok(tbl)
+}
+
+fn clone<'lua>(_: &'lua Lua, table: Table<'lua>) -> mlua::Result<Table<'lua>> {
+    Ok(table.clone())
+}
+
+fn flatten<'lua>(lua: &'lua Lua, arrays: Vec<LuaValue<'lua>>) -> mlua::Result<Vec<LuaValue<'lua>>> {
+    let mut flat_vec: Vec<LuaValue> = vec![];
+    for item in arrays {
+        match item {
+            LuaValue::Table(tbl) => {
+                let tbl_as_vec = tbl.sequence_values().filter_map(|x| x.ok()).collect();
+                let flat = flatten(lua, tbl_as_vec)?;
+                for j in flat {
+                    flat_vec.push(j);
+                }
+            }
+            LuaValue::Nil => (),
+            LuaValue::Thread(_) => (),
+            LuaValue::Error(err) => {
+                return Err(mlua::Error::external(err));
+            }
+            other => {
+                flat_vec.push(other);
+            }
+        }
+    }
+    Ok(flat_vec)
+}
+
+fn length<'lua>(_: &'lua Lua, table: Table<'lua>) -> mlua::Result<Integer> {
+    // note that # only works correctly on arrays in Lua
+    let mut len: i64 = 0;
+    for _ in table.pairs::<LuaValue, LuaValue>() {
+        len += 1
+    }
+    Ok(len)
+}
+
+fn has_key<'lua>(_: &'lua Lua, (table, key): (Table<'lua>, LuaValue)) -> mlua::Result<bool> {
+    Ok(table.contains_key(key).map_err(mlua::Error::external)?)
+}
+
+fn has_value<'lua>(_: &'lua Lua, (table, value): (Table<'lua>, LuaValue)) -> mlua::Result<bool> {
+    for pair in table.pairs::<LuaValue, LuaValue>() {
+        let (_, tbl_value) = pair.map_err(mlua::Error::external)?;
+        if tbl_value == value {
+            return Ok(true);
+        }
+    }
+    Ok(false)
+}
+
+fn to_string_fallback<'lua>(_: &'lua Lua, table: Table<'lua>) -> mlua::Result<String> {
+    Ok(format!("{:#?}", table))
+}
+
+fn to_string<'lua>(
+    lua: &'lua Lua,
+    (table, indent, skip_outer_bracket): (Table<'lua>, Option<i64>, Option<bool>),
+) -> mlua::Result<String> {
+    if let Some(ind) = indent {
+        if ind < 0 {
+            return Err(mlua::Error::external(anyhow!(
+                "Indent set to {ind}. Please use an indent ≥ 0."
+            )));
+        }
+    }
+    let result = to_string_impl(lua, (table, indent, skip_outer_bracket, 0));
+    match result {
+        Ok(res) => {
+            if skip_outer_bracket == Some(true) {
+                // we added indent too many spaces on each line
+                let extra_spaces = match indent {
+                    Some(ind) => " ".repeat(ind as usize),
+                    None => " ".repeat(2),
+                };
+                let old = ["\n", &extra_spaces].concat();
+                return Ok(res.replace(&old, "\n"));
+            }
+            Ok(res)
+        }
+        Err(err) => Err(mlua::Error::external(err)),
+    }
+}
+
+fn to_string_impl<'lua>(
```

Comment:
I also don't really like the current `to_string` implementation that much, but it is the way it is because I was trying to satisfy what I think #4328 is asking for. 

Maybe we should just remove this and use what I currently have for `to_string_fallback`? Or does the `ValuePrinter´ do anything extra compared to that?